src/libraries/System.Linq/src/System/Linq/CountBy.cs

```diff
@@ -0,0 +1,40 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        public static IEnumerable<KeyValuePair<TKey, int>> CountBy<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey>? comparer = null) where TKey : notnull
+        {
+            if (source is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
+            }
+            if (keySelector is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.keySelector);
+            }
+
+            Dictionary<TKey, int> countsBy = new(comparer);
+
+            using IEnumerator<TSource> e = source.GetEnumerator();
```

Comment:
This method is eagerly evaluating the source enumerable, returning a materialized dictionary with frequency counts. This violates the invariant of most LINQ methods returning IEnumerable where deferred evaluation is being used (this is how the related `GroupBy` works).

I wouldn't be entirely opposed to this being an eager computation, however in that case we should change the return type to be a concrete `Dictionary<TSource, int>`.

@stephentoub thoughts?

Reply:
I have added a commit to propose a lazy loading approach. Does it fit the LINQ requirements?

Reply:
> @stephentoub thoughts?

In general, I think we've kept the prompt vs lazy aspect consistent between the Xx and XxBy methods, e.g. Max and MaxBy are both prompt, Order and OrderBy are both lazy, etc. So given that `Count` is prompt, I think I'd expect `CountBy` to be prompt as well, and then as you suggest _not_ return `IEnumerable<>` but instead return an `IReadOnlyDictionary<TSource, int>` (or maybe a `long` instead of `int` if we want to avoid a future where we need `LongCountBy`).  I suggest `IReadOnlyDictionary<>` purely so that we have some flexibility, e.g. returning a singleton if the input was empty.

We should run it through API review again to discuss.

Reply:
We re-reviewed this and ended up favouring the delayed approach.